src/target/riscv/riscv-013.c

```diff
@@ -4394,182 +4394,299 @@ static int write_memory_bus_v1(struct target *target, target_addr_t address,
-				riscv_batch_add_dm_write(batch, DM_DATA0, value, false);
-				if (riscv_batch_full(batch))
-					break;
-			}
-		}
+	if (riscv_program_store(p, d, b, offset, size) != ERROR_OK)
+		return ERROR_FAIL;
 
-		result = batch_run(target, batch);
-		riscv_batch_free(batch);
-		if (result != ERROR_OK)
-			goto error;
+	if (mprven && riscv_program_csrrci(p, GDB_REGNO_ZERO, CSR_DCSR_MPRVEN,
+				GDB_REGNO_DCSR) != ERROR_OK)
+		return ERROR_FAIL;
 
-		/* Note that if the scan resulted in a Busy DMI response, it
-		 * is this read to abstractcs that will cause the dmi_busy_delay
-		 * to be incremented if necessary. */
+	return ERROR_OK;
+}
 
-		uint32_t abstractcs;
-		bool dmi_busy_encountered;
-		result = dm_op(target, &abstractcs, &dmi_busy_encountered,
-				DMI_OP_READ, DM_ABSTRACTCS, 0, false, true);
-		if (result != ERROR_OK)
-			goto error;
-		while (get_field(abstractcs, DM_ABSTRACTCS_BUSY))
-			if (dm_read(target, &abstractcs, DM_ABSTRACTCS) != ERROR_OK)
-				return ERROR_FAIL;
-		info->cmderr = get_field(abstractcs, DM_ABSTRACTCS_CMDERR);
-		if (info->cmderr == CMDERR_NONE && !dmi_busy_encountered) {
-			LOG_TARGET_DEBUG(target, "Successful (partial?) memory write");
-		} else if (info->cmderr == CMDERR_BUSY || dmi_busy_encountered) {
-			if (info->cmderr == CMDERR_BUSY)
-				LOG_TARGET_DEBUG(target, "Memory write resulted in abstract command busy response.");
-			else if (dmi_busy_encountered)
-				LOG_TARGET_DEBUG(target, "Memory write resulted in DMI busy response.");
-			riscv013_clear_abstract_error(target);
-			increase_ac_busy_delay(target);
+static int write_memory_progbuf_inner_fill_progbuf(struct target *target,
+		uint32_t size, bool mprven)
+{
+	if (riscv_save_register(target, GDB_REGNO_S0) != ERROR_OK)
+		return ERROR_FAIL;
+	if (riscv_save_register(target, GDB_REGNO_S1) != ERROR_OK)
+		return ERROR_FAIL;
 
+	struct riscv_program program;
+
+	riscv_program_init(&program, target);
+	if (riscv_program_store_mprv(&program, GDB_REGNO_S1, GDB_REGNO_S0, 0, size,
+				mprven) != ERROR_OK)
+		return ERROR_FAIL;
+
+	if (riscv_program_addi(&program, GDB_REGNO_S0, GDB_REGNO_S0, size) != ERROR_OK)
+		return ERROR_FAIL;
+
+	if (riscv_program_ebreak(&program) != ERROR_OK)
+		return ERROR_FAIL;
+
+	return riscv_program_write(&program);
+}
+
+static int write_memory_progbuf_inner(struct target *target, target_addr_t address,
+		uint32_t size, uint32_t count, const uint8_t *buffer, bool mprven)
+{
+	if (write_memory_progbuf_inner_fill_progbuf(target, size,
+				mprven) != ERROR_OK)
+		return ERROR_FAIL;
+
+	target_addr_t address_on_target = address;
+	if (write_memory_progbuf_inner_startup(target, &address_on_target, buffer, size) != ERROR_OK)
+		return ERROR_FAIL;
+
+	for (target_addr_t end_address_on_target = address + (target_addr_t)size * count,
+			next_address_on_target = address_on_target;
+			address_on_target != end_address_on_target;
+			address_on_target = next_address_on_target) {
```

Comment:
These statements inside the `for` loops are bit complex. I'd propose to take the variable definitions out.

Please, rename the variables so that the names are shorter and more descriptive. For example:

```
end_address_on_target --> end_addr  (and should be const)
address_on_target --> batch_start_addr
next_address_on_target  --> next_batch_start_addr
```

In fact, this a `do - while` loop masked as `for` loop. The assignment `next_address_on_target = address_on_target` is a "hack" to make sure the first iteration is always executed, but the assigned value is never used in the algorithm, and is over-written on line 4646.

For that reason, please, change it to `do - while`:

```suggestion
	target_addr_t batch_start_addr = address;
	const target_addr_t end_addr = address + (target_addr_t)size * count;
	target_addr_t next_batch_start_addr;
	do {
		/* ... */
		
		batch_start_addr = next_batch_start_addr;
	} while (batch_start_addr != end_addr);
```

Reply:
I have shortened the names.

As for other suggestions, you are mistaken. `&next_address_on_target` is passed to `write_memory_progbuf_try_to_write()` as `address_p` -- in/out parameter. It should be initialized in order to calculate how long the batch should be in `_fill_batch()`.

Moreover, this is not a `do - while` loop, since it is possible for this loop to have zero iterations -- one element is already written in `_startup()`.

I would also like to leave `_on_target` part in the names, since these values represent current value of `s0` register.

Reply:
Yes, I was mistaken in the two above points, sorry. Thank you for explanation.

Some of the other review suggestions should lead to further simplification of this function. So in the end, this function should become easier to understand for future readers.